### ISSUE TITLE
Subscriptions over SSE

### ISSUE THREAD

**Issue by fmatosqg** [apollographql/apollo-kotlin#3756]
**Use case**
We would like to be able to use GraphQl subscriptions over SSE (Server Sent Events) instead of Websockets.

Out of scope: the merits and demerits of SSE vs Websockets. We just want it and decided it's the best for our project ;)

**Describe the solution you'd like**

In scope: a new class implementing `SubscriptionTransport`, for example `SseSubscriptionTransport`. 

At the moment I'm reverse-engineering and reimplementing in SSE client the protocol used between client/server and implemented in `RealSubscriptionManager`. It's unclear whether our servers would want/be willing to folllow that same protocol, as described in https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md following link from the PR that first implemented it in `apollo-kotlin` aka `apollo-android` (https://github.com/apollographql/apollo-kotlin/issues/754). 

As such, it would be really useful to understand the historical origin of such protocol. Is it something born inside Apollo? I see many references to AWS AppSync but can't find the origin of such protocol. Searching in GraphQl foundation specs gave me no luck either.

**Notes**

Our company is investing time in investigating the effort and potential for subscriptions over SSE, and it's already happening internally, ETA for prouction in the next quarter (Jan-March 2022). A stretch goal would be upstreaming those changes into apollo-kotlin, preferrably v2.5.xxx and subsequent merge into v3.0.x.

**Comment by BoD** [comment]
Hi! 👋 Thanks for reaching out!

Interesting project! A few notes:
As you found out, the GraphQL specification itself is agnostic regarding the underlaying protocol to use for subscriptions, which is why other specifications have appeared over time to supplement it:
- The [subscription-transport-ws](https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md) protocol (implemented by `SubscriptionWsProtocol`) originated at Apollo and was created in collaboration with the GraphQL working group
- The [graphql-ws](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md) protocol (implemented by `GraphQLWsProtocol`) originated from @enisdenjo and aims to be standardized and become a part of GraphQL - this standardization process is happening at the GraphQL over HTTP work group [here](https://github.com/graphql/graphql-over-http/pull/140).
- The [AppSync](https://docs.aws.amazon.com/appsync/latest/devguide/real-time-websocket-client.html) protocol (implemented by `AppSyncWsProtocol`) was created at Amazon - I don't have more historical context about it

As for SSE, I notice [this project](https://github.com/enisdenjo/graphql-sse) also exists, which would probably a good basis for this work? On our side, the existence of a documented protocol is a big plus!

WebSockets / Subscriptions have been an area where we received contributions in the past and other contributions will be welcome! I can say that one big hurdle we have on this area is the difficulty to test / unit test, as servers implementing some of these protocols are not easily available.

One last note: we no longer expect to work on version 2 (except for bug fixes) however now that version 3 has been released, so it is unlikely we would want to add such a feature into version 2.

**Comment by fmatosqg** [comment]
I've had a good look at enisdenjo's SSE work, and sounds like a great starting point. Keen to adopt that.

I think my plan will be either/both mock server/ktor for dev+test purposes, on top of possibly also running node enisdenjo's for "validating" the protocol implementation.

Regarding v3, I completed migration of my preliminary POC from v2 to v3, and it's easy to notice the improvement in the amount of freedom to write the protocol there. That said, and with the substantial internal api changes, the best course of action seems to be migrate our project to v3 and in parallel develop the client SSE protocol.
However our project is quite a big one, and v2->3 migration can get nasty.

**Comment by fmatosqg** [comment]
Progress report, 
- I have a ktor server running the non-sse endpoints, and starting to write the SSE part next
- I have a `WSProtocol` and `NetworkTransport` talking to such ktor server

Unfortunately my understanding/skills into flexing `WsProtocol` into something that I need are not great. I'm thinking into making `WsProtocol` extend a new interface ( `Protocol` ) and then making myself a new (abstract?) class that extends that, like `SseProtocol`.  Because for example I have no use for a `webSocketConnection` --> `class MyProtocol :WsProtocol(webSocketConnection = UnusedWebSocketConnection()...) `

Maybe I'll also need to bend a few `ApolloClient` builders but so far I think they serve me right.

**Comment by BoD** [comment]
Hi! Thanks for the update.

What you say about having a Protocol interface makes sense, basically extracting / making it independent of the underlying "pipe". On the other hand, could the `WebSocketConnection` interface still be used for SSE (if we ignore the naming for now)? If yes it may mean `WsProtocol` could still be used as-is?

We'd be happy to talk more, and the best way is on the Kotlin Slack in the [#apollo-kotlin](https://app.slack.com/client/T09229ZC6/C01A6KM1SBZ) channel (get an invite from [here](https://slack.kotl.in/) if you're not already there).

**Comment by fmatosqg** [comment]
`WebSocketConnection` won't do (at least for now), since I need to handle 2 connections: a SSE + side channel. It's important to clarify here that I'm pursuiong the HTTP1 where all subscriptions travel through the same SSE channel (instead of the HTTP2 approach where each subscription has its own long-lived SSE channel).


# context
Side channel's purpose is to add and remove subscriptions, as well as providing the initial handshake. That's where the subscription request travels.

Long lived channel (aka SSE channel) purpose is to get subscription responses from server to client.

# problem
Having an interface like WebSocketConnection is ok-ish. But handling 2 connections through a single `object: WebSocketConnection` is very hard at best because of `suspend fun receive(): String` being used to get data from 2 different types of channel. If it were something like `suspend fun receive(): WebSocketResponse` then it would be easier to know if we are getting responses from side channel or sse channel. 

It can be done but it gets quite ugly, so I'll look into alternatives, like completely not using `WebSocketConnection` and instead handling 2 objects, one for each type of channel. 

# temp solution

```

open class SseProtocol(
   channelFactory: Channel.ChannelFactory,
) : WsProtocol(UnusedWebSocketConnection(), UnusedListener()) {

```

So far, and I don't have plans on changing this, `UnusedWebSocketConnection` and `UnusedListener` are full of `TODO`s and only purpose is to let me inherit `WsProtocol`.

`Channel.ChannelFactory` is a bit like `WebSocketEngine`, except it has 2 methods so I can make either `GET` or SSE.

**Comment by acao** [comment]
Hey so, if you're looking for a quick graphql server that supports SSE, this one is great!

https://github.com/contra/graphql-helix

**Comment by hanssonduck** [comment]
Hi there! Are there any plans to include this in the v4 release?
Here is a working multiplatform example that reads server-sent evets using Ktor https://github.com/JurajBegovac/ktor-kmm-sse.
Using Ktor for this makes sense and maybe #3796 could be done at the same time?

**Comment by BoD** [comment]
@hanssonduck This is currently not being worked on by the team and not planned to be part of v4, but we would certainly welcome a contribution for it. Note: as for #3796, having this in a dedicated module would probably be the best, if possible.

**Comment by BoD** [comment]
@hanssonduck it would definitely be a plus if this is separated in its own module

**Comment by vladimirfx** [comment]
Sorry for disturbing - does this feature is planned to implement? Unfortunately, no java/android client is exists that supports SSE transport.

**Comment by fmatosqg** [comment]
I implemented it internally in my company and had the intention of porting it here. Unfortunately our solution is quite particular to some internal decisions that won't hold as general assumptions. 

In particular the decision to hold a single sse connection for all subscriptions makes the solution 10x more complex.

**Comment by BoD** [comment]
Hey @vladimirfx, SEE is still not in our roadmap at the moment. A contribution would still be very welcome.

**Comment by bobbysothebys** [comment]
HI! Any update on this coming into the roadmap anytime soon?

**Comment by martinbonnin** [comment]
@bobbysothebys no update. This is currenly not on the raodmap but we would welcome a contribution.

**Comment by BobbyBakes** [comment]
I believe GO's standard net/http

**Comment by bobbysothebys** [comment]
I believe GO's standard net/http

**Comment by bobbysothebys** [comment]
Hi! Any hopes of this being supported in 2026?

**Comment by martinbonnin** [comment]
@wiyarmir made this work! 💙 You can now use the Avian Labs network transport to use SSE with Apollo: https://github.com/avianlabs/apollo-sse

The [docs are also updated](https://github.com/apollographql/apollo-kotlin/pull/6955) to reflect this.

_PS: I love that a pollo and Avian Labs work together, it was meant to be 🐣!_
